src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocument.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A <see cref="Document"/> that was generated by an <see cref="ISourceGenerator" />.
+    /// </summary>
+    public sealed class SourceGeneratedDocument : Document
+    {
+        internal SourceGeneratedDocument(Project project, SourceGeneratedDocumentState state)
+            : base(project, state)
+        {
+        }
+
+        private new SourceGeneratedDocumentState State => (SourceGeneratedDocumentState)base.State;
```

Comment:
i don't mind SourceGeneratedDocument as a new type.  However, do you see large value in adding the new State type?

Reply:
If we fork a compilation for a text change, we can still reuse the states for that later one if we ask for a in-progress compilation; we don't want that to root the prior Project though so the red/green node pattern still applies.